### PR TITLE
Eager Mode - Support Concatenation via aten::cat.out

### PR DESCRIPTION
**Pull request by msftlincoln**
**Description**: SignatureOnly based implementation of tensor concatenation.

**Motivation and Context**
- Why is this change required? What problem does it solve?
The update to pytorch 1.12 deprecated aten::_cat, which was previously supported but removed in that change.  This left no concatenation operator supported in eager mode, which broke downstream consumers trying to update.
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Review by jamill — DISMISSED** [review]
Looks good to me - thanks!

File: orttraining/orttraining/eager/ort_aten.cpp [hunk 1]
```diff
@@ -920,6 +920,55 @@ const at::Tensor& resize_(
...
+  }
+  auto& invoker = GetORTInvoker(tensors[0].device());
+
+  // IntArrayRef isn't writeable, convert to vector.
+  std::vector<int64_t> sizes;
+  for (auto s : tensors[0].sizes())
+    sizes.push_back(s);
+
+  // Calculate the new size of the dimension being concatenated.
+  sizes[dim] = 0;
```
Comment (WilBrady) [inline orttraining/orttraining/eager/ort_aten.cpp:948]: dim bounds check

File: orttraining/orttraining/eager/ort_aten.cpp [hunk 2]
```diff
@@ -920,6 +920,55 @@ const at::Tensor& resize_(
...
+  assert(tensors.size() > 0);
+  if (
+      std::vector<at::ScalarType> supportedTypes =
+        {at::kBFloat16, at::kBool, at::kByte, at::kDouble, at::kFloat, at::kHalf, at::kInt, at::kLong, at::kShort};
+      !IsSupportedType(tensors, supportedTypes)) {
+    return at::native::call_fallback_fn<
+      &at::native::cpu_fallback,
+      ATEN_OP(cat_out)>::call(tensors, dim, out);
+  }
+  auto& invoker = GetORTInvoker(tensors[0].device());
```
Comment (jamill) [inline orttraining/orttraining/eager/ort_aten.cpp:944]: I think the above assert on size of tensor is only run in debug builds - do we need to perform bounds checking for non-debug builds as well?

Comment (msftlincoln) [reply orttraining/orttraining/eager/ort_aten.cpp:944]: if size is zero then it'll crash on deference. it would be unexpected to concat 0 tensors together

Comment (jamill) [reply orttraining/orttraining/eager/ort_aten.cpp:944]: I was wondering if it was preferrable to throw some runtime error vs crash on dereference (or if there is some existing pattern for how this is handled in general)... I guess a similar question could be asked if the shape of the tensors is validated  before this operation. If the concat dimension is larger then the number of dimensions in one of the tensors...

File: orttraining/orttraining/eager/ort_aten.cpp [hunk 3]
```diff
@@ -920,6 +920,59 @@ const at::Tensor& resize_(
...
+
+  // IntArrayRef isn't writeable, convert to vector.
+  std::vector<int64_t> sizes;
+  for (auto s : tensors[0].sizes())
+    sizes.push_back(s);
+
+  // Calculate the new size of the dimension being concatenated.
+  sizes[dim] = 0;
+  for (auto t : tensors)
```
Comment (jamill) [inline orttraining/orttraining/eager/ort_aten.cpp:953]: From the [torch.cat](https://pytorch.org/docs/stable/generated/torch.cat.html) documentation - it looks like the input could have empty tensors? I think an empty input tensor could be an issue here.. Maybe just something to check with how default PyTorch implementation handles this case.

Comment (msftlincoln) [reply orttraining/orttraining/eager/ort_aten.cpp:953]: yes, if the first tensor is truly empty, it could be a problem. I was thinking the tensor at least needed to have the same number of dimensions, even if they were empty, but that might not be 100% accurate

Comment (jamill) [reply orttraining/orttraining/eager/ort_aten.cpp:953]: I was looking at the following lines from the documentation:

> All tensors must either have the same shape (except in the concatenating dimension) or be empty.

> Non-empty tensors provided must have the same shape, except in the cat dimension.

which I read as empty tensors might not have the same shape as the other tensors.

File: orttraining/orttraining/eager/ort_aten.cpp (hunk 2 above)
Comment (jamill) [reply orttraining/orttraining/eager/ort_aten.cpp:944]: I am fine with the current implementation - feel free to close / resolve this.

Comment (msftlincoln) [reply orttraining/orttraining/eager/ort_aten.cpp:944]: it kind-of depends.  If you call cat without an out, it'll check/normalize everything before calling out.  But if you give it an out, then it goes directly to cat.out without the checks

File: orttraining/orttraining/eager/ort_aten.cpp (hunk 3 above)
Comment (WilBrady) [reply orttraining/orttraining/eager/ort_aten.cpp:953]: What would/could be the `.sizes()` value of an empty tensor? If all the non-empty tensors had dim=3 but an empty tensor had dim=2 is that okay? I am not worried about this now but something to consider as we move eager forward.

**Review by WilBrady — APPROVED** [review]
